internal/allocator/allocator.go

```diff
@@ -194,113 +222,116 @@ func (a *Allocator) Unassign(svc string) bool {
 
 	al := a.allocated[svc]
 	delete(a.allocated, svc)
-	for _, port := range al.ports {
-		if curSvc := a.portsInUse[al.ip.String()][port]; curSvc != svc {
-			panic(fmt.Sprintf("incoherent state, I thought port %q belonged to service %q, but it seems to belong to %q", port, svc, curSvc))
+	for _, ip := range al.ips {
+		for _, port := range al.ports {
+			if curSvc := a.portsInUse[ip.String()][port]; curSvc != svc {
+				panic(fmt.Sprintf("incoherent state, I thought port %q belonged to service %q, but it seems to belong to %q", port, svc, curSvc))
+			}
+			delete(a.portsInUse[ip.String()], port)
+		}
+
+		delete(a.servicesOnIP[ip.String()], svc)
+		if len(a.portsInUse[ip.String()]) == 0 {
+			delete(a.portsInUse, ip.String())
+			delete(a.sharingKeyForIP, ip.String())
+		}
+		a.poolIPsInUse[al.pool][ip.String()]--
+		if a.poolIPsInUse[al.pool][ip.String()] == 0 {
+			// Explicitly delete unused IPs from the pool, so that len()
+			// is an accurate count of IPs in use.
+			delete(a.poolIPsInUse[al.pool], ip.String())
 		}
-		delete(a.portsInUse[al.ip.String()], port)
-	}
-	delete(a.servicesOnIP[al.ip.String()], svc)
-	if len(a.portsInUse[al.ip.String()]) == 0 {
-		delete(a.portsInUse, al.ip.String())
-		delete(a.sharingKeyForIP, al.ip.String())
-	}
-	a.poolIPsInUse[al.pool][al.ip.String()]--
-	if a.poolIPsInUse[al.pool][al.ip.String()] == 0 {
-		// Explicitly delete unused IPs from the pool, so that len()
-		// is an accurate count of IPs in use.
-		delete(a.poolIPsInUse[al.pool], al.ip.String())
 	}
 	stats.poolActive.WithLabelValues(al.pool).Set(float64(len(a.poolIPsInUse[al.pool])))
 	return true
 }
 
-func cidrIsIPv6(cidr *net.IPNet) bool {
-	return cidr.IP.To4() == nil
-}
-func ipIsIPv6(ip net.IP) bool {
-	return ip.To4() == nil
-}
-
 // AllocateFromPool assigns an available IP from pool to service.
-func (a *Allocator) AllocateFromPool(svc string, isIPv6 bool, poolName string, ports []Port, sharingKey, backendKey string) (net.IP, error) {
+func (a *Allocator) AllocateFromPool(svc string, ipFamily IPFamily, poolName string, ports []Port, sharingKey, backendKey string) ([]net.IP, error) {
 	if alloc := a.allocated[svc]; alloc != nil {
-		// Handle the case where the svc has already been assigned an IP but from the wrong family.
-		// This "should-not-happen" since the "ipFamily" is an immutable field in services.
-		if isIPv6 != ipIsIPv6(alloc.ip) {
-			return nil, fmt.Errorf("IP for wrong family assigned %s", alloc.ip.String())
+		if ipFamily != DualStack {
+			// Handle the case where the svc has already been assigned an IP but from the wrong family.
+			// This "should-not-happen" since the "ipFamily" is an immutable field in services.
+			if ipFamily != getIPFamily(alloc.ips[0]) {
```

Comment:
I think this must check ipFamily of `alloc.ips` compared to the ipFamily parameter.